File: public/api/crops.php
<?php

require dirname(__DIR__, 2) . '/vendor/autoload.php';

use App\Database;
use App\ImageRepository;
use App\CropRepository;
use App\PresetLoader;
use App\AppConfig;
use App\ComponentLoader;

// Auth::requireAuth();

header('Content-Type: application/json');

$method   = $_SERVER['REQUEST_METHOD'];
$db       = Database::get();
$images   = new ImageRepository($db);
$crops    = new CropRepository($db);

if ($method === 'GET') {
    $imageId     = $_GET['image_id']     ?? '';
    $componentId = $_GET['component_id'] ?? '';

    if ($imageId === '') {
        http_response_code(400);
        echo json_encode(['error' => 'Missing image_id']);
        exit;
    }

    $image = $images->findById($imageId);
    if ($image === null) {
        http_response_code(404);
        echo json_encode(['error' => 'Image not found']);
        exit;
    }

    $savedCrops = [];
    foreach ($crops->forImage($imageId) as $row) {
        $savedCrops[$row['preset_id'] . ':' . $row['variant_id']] = $row;
    }

    $outputFolder  = AppConfig::outputFolder();
    $defaultFormat = AppConfig::defaultFormat();
    $result        = [];

    // If a component_id is given, only show that component's presets
    $allPresets = PresetLoader::all();
    if ($componentId !== '') {
        $component = ComponentLoader::findComponent($componentId);
        if ($component !== null) {
            $allowedPresetIds = $component['presets'] ?? [];
            $allPresets = array_values(array_filter($allPresets, fn($p) => in_array($p['id'], $allowedPresetIds, true)));
        }
    }

    foreach ($allPresets as $preset) {
        $presetFormat  = $preset['format'] ?? $defaultFormat;
        $variantStates = [];

        foreach ($preset['variants'] as $variant) {
            $key    = $preset['id'] . ':' . $variant['id'];
            $saved  = $savedCrops[$key] ?? null;
            $format = $variant['format'] ?? $presetFormat;
            $ext    = $format;
            $base   = $componentId !== ''
                ? $componentId . '_' . $variant['id']
                : $imageId . '_' . $preset['id'] . '_' . $variant['id'];

            $file1x = $base . '_1x.' . $ext;
            $file2x = $base . '_2x.' . $ext;

            $variantStates[] = [
                'id'           => $variant['id'],
                'label'        => $variant['label'],
                'width'        => $variant['width'],
                'height'       => $variant['height'],
                'format'       => $format,
                'crop'         => $saved ? [
                    'x' => (int) $saved['crop_x'],
                    'y' => (int) $saved['crop_y'],
                    'w' => (int) $saved['crop_w'],
                    'h' => (int) $saved['crop_h'],
                ] : null,
                'generated_at' => $saved['generated_at'] ?? null,
                'files'        => [
                    '1x' => $file1x,
                    '2x' => $file2x,
                ],
                'files_exist'  => $outputFolder !== '' && file_exists($outputFolder . '/' . $file1x),
            ];
        }

        $result[] = [
            'id'       => $preset['id'],
            'label'    => $preset['label'],
            'variants' => $variantStates,
        ];
    }

    echo json_encode(['presets' => $result]);
    exit;
}

if ($method === 'POST') {
    $body = json_decode(file_get_contents('php://input'), true);

    $imageId   = $body['image_id']   ?? '';
    $presetId  = $body['preset_id']  ?? '';
    $variantId = $body['variant_id'] ?? '';
    $x         = (int) ($body['crop_x'] ?? 0);
    $y         = (int) ($body['crop_y'] ?? 0);
    $w         = (int) ($body['crop_w'] ?? 0);
    $h         = (int) ($body['crop_h'] ?? 0);

    if ($imageId === '' || $presetId === '' || $variantId === '') {
        http_response_code(400);
        echo json_encode(['error' => 'Missing required fields']);
        exit;
    }

    if ($images->findById($imageId) === null) {
        http_response_code(404);
        echo json_encode(['error' => 'Image not found']);
        exit;
    }

    if (PresetLoader::findVariant($presetId, $variantId) === null) {
        http_response_code(400);
        echo json_encode(['error' => 'Unknown preset/variant']);
        exit;
    }

    if ($w <= 0 || $h <= 0) {
        http_response_code(422);
        echo json_encode(['error' => 'Invalid crop dimensions']);
        exit;
    }

    $crops->upsert($imageId, $presetId, $variantId, $x, $y, $w, $h);

    echo json_encode(['success' => true]);
    exit;
}

http_response_code(405);
echo json_encode(['error' => 'Method not allowed']);

File: internal/handler/crops.go
package handler

import (
	"imagemanager/internal/config"
	"imagemanager/internal/model"
	"imagemanager/internal/repo"
	"net/http"
	"os"
	"path/filepath"
)

func (a *App) Crops(w http.ResponseWriter, r *http.Request) {
	imgRepo := repo.NewImageRepo(a.DB)
	cropRepo := repo.NewCropRepo(a.DB)

	switch r.Method {
	case http.MethodGet:
		imageID := r.URL.Query().Get("image_id")
		componentID := r.URL.Query().Get("component_id")

		if imageID == "" {
			jsonErr(w, 400, "Missing image_id")
			return
		}

		image, err := imgRepo.FindByID(imageID)
		if err != nil || image == nil {
			jsonErr(w, 404, "Image not found")
			return
		}

		savedCrops, _ := cropRepo.ForImage(imageID)
		cropMap := make(map[string]map[string]any)
		for _, c := range savedCrops {
			key := c.PresetID + ":" + c.VariantID
			cropMap[key] = map[string]any{
				"crop_x":       c.CropX,
				"crop_y":       c.CropY,
				"crop_w":       c.CropW,
				"crop_h":       c.CropH,
				"generated_at": c.GeneratedAt,
			}
		}

		cfg, _ := config.LoadAppConfig(a.AppRoot)
		allPresets, _ := config.LoadPresets(a.AppRoot)

		// Filter by component if provided
		if componentID != "" {
			components, _ := config.LoadComponents(a.AppRoot)
			comp := config.FindComponent(components, componentID)
			if comp != nil {
				filtered := make([]model.Preset, 0)
				for _, p := range allPresets {
					for _, pid := range comp.Presets {
						if p.ID == pid {
							filtered = append(filtered, p)
							break
						}
					}
				}
				allPresets = filtered
			}
		}

		result := make([]any, 0)
		for _, preset := range allPresets {
			presetFormat := preset.Format
			if presetFormat == "" {
				presetFormat = cfg.DefaultFormat
			}

			variants := make([]any, 0)
			for _, variant := range preset.Variants {
				key := preset.ID + ":" + variant.ID
				saved := cropMap[key]

				format := variant.Format
				if format == "" {
					format = presetFormat
				}
				ext := format

				base := componentID + "_" + variant.ID
				if componentID == "" {
					base = imageID + "_" + preset.ID + "_" + variant.ID
				}
				file1x := base + "_1x." + ext
				file2x := base + "_2x." + ext

				filesExist := false
				if cfg.OutputFolder != "" {
					_, err := os.Stat(filepath.Join(cfg.OutputFolder, file1x))
					filesExist = err == nil
				}

				var cropObj any
				var generatedAt any
				if saved != nil {
					cropObj = map[string]int{
						"x": saved["crop_x"].(int),
						"y": saved["crop_y"].(int),
						"w": saved["crop_w"].(int),
						"h": saved["crop_h"].(int),
					}
					generatedAt = saved["generated_at"]
				}

				variants = append(variants, map[string]any{
					"id":           variant.ID,
					"label":        variant.Label,
					"width":        variant.Width,
					"height":       variant.Height,
					"format":       format,
					"crop":         cropObj,
					"generated_at": generatedAt,
					"files": map[string]string{
						"1x": file1x,
						"2x": file2x,
					},
					"files_exist": filesExist,
				})
			}

			result = append(result, map[string]any{
				"id":       preset.ID,
				"label":    preset.Label,
				"variants": variants,
			})
		}

		jsonOK(w, map[string]any{"presets": result})

	case http.MethodPost:
		var body struct {
			ImageID   string `json:"image_id"`
			PresetID  string `json:"preset_id"`
			VariantID string `json:"variant_id"`
			CropX     int    `json:"crop_x"`
			CropY     int    `json:"crop_y"`
			CropW     int    `json:"crop_w"`
			CropH     int    `json:"crop_h"`
		}
		if err := readJSON(r, &body); err != nil {
			jsonErr(w, 400, "Invalid request body")
			return
		}

		if body.ImageID == "" || body.PresetID == "" || body.VariantID == "" {
			jsonErr(w, 400, "Missing required fields")
			return
		}

		if img, _ := imgRepo.FindByID(body.ImageID); img == nil {
			jsonErr(w, 404, "Image not found")
			return
		}

		presets, _ := config.LoadPresets(a.AppRoot)
		preset := config.FindPreset(presets, body.PresetID)
		if preset == nil || config.FindVariant(preset, body.VariantID) == nil {
			jsonErr(w, 400, "Unknown preset/variant")
			return
		}

		if body.CropW <= 0 || body.CropH <= 0 {
			jsonErr(w, 422, "Invalid crop dimensions")
			return
		}

		if err := cropRepo.Upsert(body.ImageID, body.PresetID, body.VariantID, body.CropX, body.CropY, body.CropW, body.CropH); err != nil {
			jsonErr(w, 500, "Failed to save crop")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	default:
		jsonErr(w, 405, "Method not allowed")
	}
}
